### ISSUE TITLE
[IDEA] Add oEmbed / OpenGraph metatags

### ISSUE THREAD

**Issue by twMat** [TiddlyWiki/TiddlyWiki5#8433]
A low hanging fruit, I think!

I want to be able to post a link to a TW and have it show up with a "preview", as is the case for many links to websites, on many boards etc. For example, on talks.tiddlywiki.com this works for links to "CSS tricks" articles but for links to a TW you get this sadness:

<img width="464" alt="image" src="https://user-images.githubusercontent.com/4534056/194725561-bb9a54c7-2e67-48d7-b8bb-0dec9a647e49.png">

...and so the user is forced to type `[label](url)` and it only shows the plain blue label. My proposal would make it possible to just directly paste the url and it'd automatically get a nice preview.

I posted a [request](https://github.com/simonbaird/tiddlyhost/issues/250) to @simonbaird / Tiddlyhost on this and he made me understand that it is a TW matter. It seems it would be pretty simple to solve though!

**Proposal**
Add metatags for this  - or if thats in inappropriate, then add some setting to activate it. It is not feasible to have users manually add this.

**Comment by twMat** [comment]
...by the way, I think this definitely aligns with the objective to "market TW". A rich link like that should be significantly more attractive to click than just a plain prettylink, not to mention an ugly url.

**Comment by pmario** [comment]
>A low hanging fruit, I think!

Not really simple, since users will need to add a unique description field to every tiddler, so the OG meta data can show it. To create a usable box, users would also need an image url for every tiddler. Single file wikis are not well suited to deal with images.
Having 1 description and 1 image URL per wiki, the boxes will be very boring, since they will look the same for every tiddler-link. 

The following data will need to be different for every permalink, to create a usable box.

```
<meta property="og:title" content="Colby Fayock - A UX Designer &amp; Front-end Developer Blog" />
<meta property="og:type" content="website" />
<meta property="og:image" content="/static/website-social-card-44070c4a901df708aa1563ac4bbe595a.jpg" />
<meta property="og:url" content="https://www.colbyfayock.com" />
```

I personally include links to "support" the content I'm writing. I do not intend to create a distraction that leads readers away from my posts. I want that they read my content till the end. That's why sometimes, I do add external links at the end of a post. 

I actually find those boxes annoying even if they show something. I think they interrupt the flow of reading. It's no wonder that Twitter promotes it. They only had 140 characters, that are not worth reading. So they need a fancy link-box, with content, that was created somewhere else.

**Comment by Jermolene** [comment]
Thanks @twMat 

As far as I can tell, it would not be possible to implement oEmbed for single file TiddlyWiki's. The [oEmbed specification](https://oembed.com) makes it clear that the server must be able to respond to API calls.

Meanwhile, [the OpenGraph specification](https://ogp.me) just defines metadata tags to be used within an HTML page, and so is compatible with single file TiddlyWiki. The meta tags can be added with `$:/tags/RawMarkup` or `$:/tags/RawMarkupWikified`.

The potential problem is with the `og:image` property which specifies the image to be displayed in the preview. Apparently, it requires the full URL of an image, and does not accept base 64 data URIs. If so, that would be very disappointing as it would require users to save their chosen preview image as a separate file somewhere.

Anyhow, I don't think any core modifications are needed in order to be able to experiment with this stuff.

**Comment by twMat** [comment]
@pmario 

>Having 1 description and 1 image URL per wiki, the boxes will be very boring, since they will look the same for every tiddler-link.

The description is the tiddler title (possibly a transcluded description if the tiddler has one) and the image is the wiki favion. (Saying that the favicon is boring is like saying avatars are boring.)

>I personally ... do not ... leads readers away from my posts. I want that they read my content till the end. 

But currently, on talk.tiddlywiki.com, one *does* automatically get the "rich link" when you paste a raw url. It just looks bad. So you **have** to make a pretty link. Lucky, if that is what you *want*, as you evidently do!. But what if you *don't* want that?

**Comment by twMat** [comment]
@Jermolene 

>As far as I can tell, it would not be possible to implement oEmbed for single file TiddlyWiki's. The [oEmbed specification](https://oembed.com/) makes it clear that the server must be able to respond to API calls.

If I understand @simonbaird right, he does succeed in doing it for Tiddlyhost as he [show here](https://github.com/simonbaird/tiddlyhost/issues/250#issuecomment-1273225172). Are TWs converted into nodejs versions after upload? (because users do upload single file TWs) or is perhaps the example Simon shows not a single file?

>The potential problem is with the og:image property which specifies the image to be displayed in the preview.

Daniel did set up TW so each wiki has an addressable little image. My guess is that a big number of the community sets up their public wikis via Tiddlyhost.

**Comment by Jermolene** [comment]
Hi @twMat what @simonbaird has done is very cool, there's now a preview image automatically created for each wiki that can be used as the OpenGraph image. That's awesome, but it's very much specific to TiddlyHost.

**Comment by simonbaird** [comment]
> ...and does not accept base 64 data URIs...

That's too bad. If base64 urls worked, then I think this would be a lot more doable.

**Comment by twMat** [comment]
For Let me rephrase my question a bit:

Can TW be made to "sense" if it was served from a server that "responds to API calls" and in that case use this fact for this context?

And for fFor og:image this sounds like it *could* work to [Output the first image url of a specific div for og:image tag?](https://stackoverflow.com/questions/21651134/output-the-first-image-url-of-a-specific-div-for-ogimage-tag) ...but maybe [not](https://stackoverflow.com/questions/47268196/can-facebook-ogimage-dynamically-pull-image-from-a-specific-div), if these two posts concern the same thing (I'm not qualified to tell).

**Comment by Jermolene** [comment]
> That's too bad. If base64 urls worked, then I think this idea would be a lot more feasible.

It would still be very terrific to get it working on TiddlyHost, and if there's any core plumbing that will make that easier I'm very happy to consider it.

**Comment by simonbaird** [comment]
> > That's too bad. If base64 urls worked, then I think this idea would be a lot more feasible.
> 
> It would still be very terrific to get it working on TiddlyHost, and if there's any core plumbing that will make that easier I'm very happy to consider it.

Perhaps TiddlyWiki could provide macro that produces something similar to https://simon.tiddlyhost.com/#OpenGraph%20Tags%20Test . The title and description could be pulled in from from the appropriate title and subtitle tiddlers. I guess the one thing that would need to be provided by the user would be a url for the preview image.

**Comment by simonbaird** [comment]
FWIW I'm not vouching for the specific markup created by https://www.opengraph.xyz , but I expect it's a good starting point.

**Comment by Jermolene** [comment]
> Perhaps TiddlyWiki could provide a macro that produces something similar to https://simon.tiddlyhost.com/#OpenGraph%20Tags%20Test . The title and description could be pulled in from from the appropriate title and subtitle tiddlers. I guess the one thing that would need to be provided by the user would be a url for the preview image.

It probably makes sense to make the entire UI be part of the core. It would be useful on tiddlywiki.com, and that would make a good reference implementation.

The procedure would need to be used in a tiddler tagged $:/tags/RawMarkupWikified to allow the transclusion of user fields.

**Comment by Jermolene** [comment]
Hi @simonbaird I've made an initial PR at #8441. The idea would be that TiddlyHost would inject the tiddlers `$:/SiteDomain`, `$:/SiteThumbnailUrl` and `$:/SiteUrl`.
